### PR TITLE
[Inductor] Retain default reduction config in max autotune

### LINKED ISSUE #194736: Inductor: retain the default reduction config in max-autotune
### 🐛 Describe the bug

The max-autotune reduction set can omit the configuration chosen by the default persistent-reduction heuristic. More tuning can then exclude the best known candidate.
Max-autotune should start with the exact default candidate. It can then append its wider search set and remove duplicates.

## Motivation

FP16 row mean with 2,048 rows reproduces the gap at three reduction widths:

| Width | Default config | Default | Best max-only candidate | Advantage |
| ---: | --- | ---: | ---: | ---: |
| 256 | X4, one warp | 4.160 us | 4.288 us | 1.031x |
| 512 | X2, one warp | 5.504 us | 5.856 us | 1.064x |
| 1024 | X1, one warp | 4.832 us | 5.856 us | 1.212x |

Current source builds persistent-reduction configurations here:
[`triton_heuristics.py`](https://github.com/pytorch/pytorch/blob/b1716d913ad13513f6184bad3d025a5b259329f3/torch/_inductor/heuristics/triton_codegen/reduction.py#L416-L498).

## Reproducer

```python
import torch
import triton


def f(x):
    return x.mean(dim=-1)


for width in (256, 512, 1024):
    x = torch.randn(2048, width, device="cuda", dtype=torch.float16)
    for mode in (None, "max-autotune-no-cudagraphs"):
        compiled = torch.compile(f, fullgraph=True, mode=mode)
        torch.testing.assert_close(compiled(x), f(x))
        for _ in range(100):
            compiled(x)
        print(width, mode, triton.testing.do_bench(lambda: compiled(x)))
```

Use `TORCH_LOGS=output_code` to compare the selected launch configurations.

### Error logs

_No response_

### Versions

Latest main (b1716d913ad13513f6184bad3d025a5b259329f3)

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by shino16**
Fixes #194736

This keeps the exact default persistent-reduction config in the max-autotune search, then appends the existing backend-specific candidates with stable de-duplication for non-tiled reductions. Max autotune therefore remains a superset of the default search instead of possibly omitting its selected config.

The issue's FP16 row-mean examples reported the omitted config as 3% to 21% faster. On GB200, the restored candidates were correct and spill-free but did not win, so this fixes search-space completeness rather than claiming a universal speedup.

## Test plan

-   python test/inductor/test_triton_heuristics.py -k persistent_max_autotune -k rocm_persistent_tiled_max_autotune
- The tests cover reduction widths 256, 512, and 1,024, checking exact candidate order, retained max-autotune candidates, no duplicates, and unchanged tiled ROCm configs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_triton_heuristics.py [hunk 1]
```diff
@@ -209,6 +212,67 @@ def test_native_matmul_config_block_numel_limit(self):
...
+            ],
+            expected_max_configs[rnumel],
+        )
+
+        default_keys = [triton_config_to_hashable(c) for c in default_configs]
+        max_autotune_keys = [triton_config_to_hashable(c) for c in max_autotune_configs]
+        self.assertEqual(max_autotune_keys[: len(default_keys)], default_keys)
+        self.assertEqual(len(max_autotune_keys), len(set(max_autotune_keys)))
+
+    def test_rocm_persistent_tiled_max_autotune_configs_remain_tiled(self):
```
Comment (shino16) [inline test/inductor/test_triton_heuristics.py:344]: This test protects the `"y" not in size_hints` guard. Without it, ROCm appends a non-tiled max-autotune config without `YBLOCK` to tiled reductions. The test ensures tiled ROCm configs remain unchanged and valid.

File: test/inductor/test_triton_heuristics.py [hunk 2]
```diff
@@ -209,6 +212,67 @@ def test_native_matmul_config_block_numel_limit(self):
         with self.assertRaisesRegex(AssertionError, "exceeds Triton maximum"):
             make_matmul_triton_config({"x": 256, "y": 128, "r": 64}, 8, 1)
 
+    @parametrize("rnumel", [256, 512, 1024])
+    def test_persistent_max_autotune_includes_default_config(self, rnumel):
```
Comment (IvanYashchuk) [inline test/inductor/test_triton_heuristics.py]: Do other tests cover `RSPLIT_SIZE`, `OUTER`/`OUTER_TINY`, and `max_autotune_pointwise` cases, which this PR touches?

**Review by IvanYashchuk — APPROVED** [review]
I think the implementation is right: it computes the existing default candidate first, keeps it at the front of the non-tiled max-autotune set, and removes exact duplicates.

The remaining gap for me is test coverage across the other branches this PR changes. Let's merge once the INNER boundary/RSPLIT, OUTER/OUTER_TINY, and both max-autotune flag paths are verified to be covered with tests.

File: test/inductor/test_triton_heuristics.py (hunk 2 above)
Comment (shino16) [reply test/inductor/test_triton_heuristics.py]: Thanks, I added some parametrization to cover those cases.

**Review by eellison — APPROVED** [review]
@claude double check that the reduction heuristics behave the same as before except in max autotune

**Review by sylvesterkaczmarek — COMMENTED** [review]
Keeping the existing default candidate first and treating max-autotune configs as an additive superset is the right invariant. Stable dedup after combining the default path with the extra search space avoids both duplicate launches and accidental loss of the known-good configuration.

**Comment by shino16** [comment]
@eellison I asked GPT-6 Astra to do the check requested above on behalf. 

> Confirmed on 8e193d53c2bd: reduction heuristics behave the same as before when max-autotune is disabled. I executed the base and head persistent-reduction heuristics across 18,432 cases covering CUDA, ROCm, and XPU, every reduction hint, tiled and non-tiled shapes, RSPLIT settings, batch-invariant mode, and both omitted and explicitly disabled max-autotune flags. Every case returned identical configuration fields in identical order.
> 
> The source diff also confirms that non-persistent heuristics and native-matmul early returns are unchanged. The candidate union and de-duplication are gated on max-autotune being enabled and the reduction being non-tiled.
